Implement builtin:: functions using ops

leonerd opened the pull request:
* Sets a callchecker to rewrite callsites to these new builtin functions into optrees
 * Sets prototypes on the functions - `""` for the constant ones, `"$"` for the unary function
 * Adds new opcode `OP_ISBOOL`
 * Existing XSUBs still need to remain because of cases like `$code = \&builtin::true`
 * `true` and `false` become OP_CONST and thus will const-fold as per normal

ilmari commented on builtin.c:
```diff
@@ -34,20 +41,125 @@ XS(XS_builtin_false)
+
+    SV *constval;
+    switch(builtin->ckval) {
+        case 0: constval = &PL_sv_no; break;
+        case 1: constval = &PL_sv_yes; break;
```
Do we want an `enum` for these, like we have for the functions?

leonerd replied:
Possibly; if we create any more. I didn't imagine we'd need more than just these two constants at the moment though.

ilmari replied:
Might as well do it straight away, it doesn't cost anything. And there's only one function, and that still/already has an `enum`.

leonerd replied:
Fair; added 65ec621870

ilmari commented on builtin.c:
```diff
@@ -207,11 +212,18 @@ Perl_boot_core_builtin(pTHX)
 
         if(builtin->checker) {
             cv_set_call_checker_flags(cv, builtin->checker, newSVuv(PTR2UV(builtin)), 0);
+
```
Spurious blank line.

leonerd replied:
Thanks; fixed 8e443f3d86

ilmari commented on builtin.c:
```diff
@@ -91,9 +208,23 @@ XS(XS_builtin_import)
-    newXS_flags("builtin::isbool", &XS_builtin_isbool, __FILE__, NULL, 0);
+    I32 i;
+    for(i = 0; builtins[i].name; i++) {
+        const struct BuiltinFuncDescriptor *builtin = builtins + i;
```
Given that `builtins` is declared as an array, and we're using `builtins[i].name` above, `builtins[i]` would be more consistent.

leonerd replied:
Fixed.

ilmari reviewed (changes requested):
How about some tests in `t/perf/optree.t` and `lib/B/Deparse.t`?

leonerd committed:
Direct optree implementations of builtin:: functions

Turn builtin::true/false into OP_CONSTs

Add a dedicated OP_ISBOOL, make an efficient op version of builtin::isbool()

leonerd commented:
@ilmari I've added a test in `lib/B/Deparse.t` now. I'll take a look at the optree one.

leonerd commented:
After a conversation with haarg on IRC I ended up changing the implementation, to just add an individual opcode for `isbool`, rather than the indirect dispatch approach. This is likely nicer all round.

leonerd commented:
Various tests added; reviewers PTAL.
